src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,260 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+
+#include "absl/container/btree_map.h"
+#include "absl/strings/str_format.h"
+#include "ray/common/grpc_util.h"
+#include "ray/rpc/client_call.h"
+#include "ray/rpc/grpc_client.h"
+
+namespace ray {
+namespace rpc {
+
+// Define a void retryable RPC client method.
+#define VOID_RETRYABLE_RPC_CLIENT_METHOD(                                        \
+    retryable_rpc_client, SERVICE, METHOD, rpc_client, method_timeout_ms, SPECS) \
+  void METHOD(const METHOD##Request &request,                                    \
+              const ClientCallback<METHOD##Reply> &callback) SPECS {             \
+    retryable_rpc_client->CallMethod<SERVICE, METHOD##Request, METHOD##Reply>(   \
+        &SERVICE::Stub::PrepareAsync##METHOD,                                    \
+        rpc_client,                                                              \
+        #SERVICE ".grpc_client." #METHOD,                                        \
+        request,                                                                 \
+        callback,                                                                \
+        method_timeout_ms);                                                      \
+  }
+
+/**
+ * The client makes RPC calls through the provided underlying grpc client.
+ * If the call goes through, the user provided callback is invoked.
+ * If the call fails due to transient network error, it is added to a retry queue.
+ * The client waits for the grpc channel reconnection to resend the requests.
+ * If the total number of request bytes in the queue exceeds max_pending_requests_bytes,
+ * the io context thread is blocked until some requests are resent.
+ * If a call's timeout_ms reaches during retry, its callback is called with
+ * Status::TimedOut. If the whole client does not reconnect within
+ * server_unavailable_timeout_seconds, server_unavailable_timeout_callback is invoked.
+ *
+ * When all callers of the client release the shared_ptr of the client, the client
+ * destructor is called and the client is shut down.
+ */
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
```

Comment:
Curious why enabled shared from this is needed here?

Reply:
NVM, I just found it's used in request.

Reply:
I called `enabled_from_this()`.